crates/wasmtime/src/module/registry.rs

```diff
@@ -15,30 +16,54 @@ lazy_static::lazy_static! {
 
 /// Used for registering modules with a store.
 ///
-/// The map is from the ending (exclusive) address for the module code to
-/// the registered module.
-///
-/// The `BTreeMap` is used to quickly locate a module based on a program counter value.
+/// Note that the primary reason for this registry is to ensure that everything
+/// in `Module` is kept alive for the duration of a `Store`. At this time we
+/// need "basically everything" within a `Moudle` to stay alive once it's
+/// instantiated within a store. While there's some smaller portions that could
+/// theoretically be omitted as they're not needed by the store they're
+/// currently small enough to not worry much about.
 #[derive(Default)]
 pub struct ModuleRegistry {
-    modules_with_code: BTreeMap<usize, Arc<RegisteredModule>>,
-    modules_without_code: Vec<Arc<CompiledModule>>,
+    // Sorted by module start address in memory, used for looking up based on a
+    // pc of a wasm function.
+    modules_with_code: Vec<Module>,
```

Comment:
Why switch `modules_with_code` from a `BTreeMap` to a sorted `Vec`? This is introducing  `O(n)` runtime to registering new modules. Is the binary search in the sorted `Vec` really that much faster that we are willing to take the hit on `O(n)` module registration?

Reply:
Does the `BTreeMap` actually have overhead for a small number of modules? If they all fit in a single node, then I think there wouldn't be any overhead at all?

Without profiling evidence to the contrary, I'd lean towards keeping the `BTreeMap` and avoiding turning registration from `O(log n)` to `O(n)`. Especially since, with the component model, we will expect smaller but more modules to become the norm.